### PR TITLE
feature/672 address codesmells

### PR DESCRIPTION
**Pull request by cschwinderg**
## Related Issues:
* [HMW-672](https://jira.epa.gov/browse/HMW-672)

## Main Changes:
* Addressed codesmells flagged by sonarcloud. I definitely didn't get them all, but hopefully this will significantly reduce the list.
* Did another round of dependency updates, since it has been a while.

## Steps To Test:
1. Test as many features as possible, to make sure nothing broke

## TODO:
- [ ] Re-review sonarcloud, github dependabot and snyk after this is merged.

### CODE REVIEW

File: app/client/src/components/pages/StateTribal.js [hunk 1]
```diff
@@ -513,21 +518,28 @@ function StateTribal() {
...
-                          key={`source-key-${sourceIndex}`}
-                          onClick={() => {
-                            setSelectedSource(source);
-                            setSourcesVisible(false);
-
-                            const searchInput =
-                              document.getElementById('hmw-search-input');
-                            if (searchInput) searchInput.focus();
-                          }}
+                          key={uuid()}
```
Comment (maxdiebold-erg) [inline app/client/src/components/pages/StateTribal.js:540]: I think the source could just be used here:
```suggestion
                          key={source}
```

File: app/client/src/components/pages/StateTribal.js [hunk 2]
```diff
@@ -612,38 +624,35 @@ function StateTribal() {
...
+                                  value = metric.value;
+                                } else if (value <= 1) {
+                                  // numbers <=1 convert to percentages
+                                  value = (value * 100).toLocaleString() + '%';
+                                } else {
+                                  value = value.toLocaleString();
+                                }
+
+                                return (
+                                  <div css={keyMetricStyles} key={uuid()}>
```
Comment (maxdiebold-erg) [inline app/client/src/components/pages/StateTribal.js:644]: Described above:
```suggestion
                                  <div css={keyMetricStyles} key={simpleHash(metric.label)}>
```

File: app/client/src/components/shared/LocationSearch.js [hunk 3]
```diff
@@ -581,60 +626,16 @@ function LocationSearch({ route, label }: Props) {
...
-                  } else if (source.source.name === 'Watersheds') {
-                    // extract the huc from "Watershed (huc)" and search on the huc
-                    const huc = result.text.split('(')[1].replace(')', '');
-                    formSubmit(huc);
+                onClick={handleSuggestionClick}
+                onKeyDown={handleSuggestionClick}
+              >
+                {getHighlightParts(result.text, inputText).map((part) => {
+                  if (part.toLowerCase() === inputText.toLowerCase()) {
+                    return <strong key={uuid()}>{part}</strong>;
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/LocationSearch.js:634]: I think this is fine as is, honestly the index would be best here.

File: app/client/src/components/shared/LocationSearch.js [hunk 4]
```diff
@@ -581,60 +626,16 @@ function LocationSearch({ route, label }: Props) {
...
-                            res.features[0].geometry.centroid ??
-                            res.features[0].geometry;
-                          formSubmit(result.text, center);
-                          searchWidget.search(result.text);
-                        }
-                      })
-                      .catch((err) => {
-                        setErrorMessage(webServiceErrorMessage);
-                      });
+                    return <Fragment key={uuid()}>{part}</Fragment>;
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/LocationSearch.js:636]: Fine as is, or revert to the index.

File: app/client/src/components/shared/WaterSystemSummary.js [hunk 5]
```diff
@@ -422,7 +423,7 @@ function WaterSystemSummary({ state }: Props) {
                 return (
                   <ul css={rowsStyles}>
                     {labels.map((label, index) => (
-                      <li css={rowStyles} key={index}>
+                      <li css={rowStyles} key={uuid()}>
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/WaterSystemSummary.js:426]: Described above:
```suggestion
                      <li css={rowStyles} key={simpleHash(label)}>
```

File: app/client/src/components/shared/WaterSystemSummary.js [hunk 6]
```diff
@@ -435,15 +436,15 @@ function WaterSystemSummary({ state }: Props) {
                   <table css={tableStyles} className="table">
                     <thead>
                       <tr>
-                        {labels.map((item, index) => (
-                          <th key={index}>{item}</th>
+                        {labels.map((item) => (
+                          <th key={uuid()}>{item}</th>
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/WaterSystemSummary.js:440]: Described above:
```suggestion
                          <th key={simpleHash(item)}>{item}</th>
```

File: app/client/src/components/shared/WaterSystemSummary.js [hunk 7]
```diff
@@ -435,15 +436,15 @@ function WaterSystemSummary({ state }: Props) {
...
                         ))}
                       </tr>
                     </thead>
                     <tbody>
                       <tr>
-                        {values.map((item, index) => (
-                          <td key={index}>{item}</td>
+                        {values.map((item) => (
+                          <td key={uuid()}>{item}</td>
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/WaterSystemSummary.js:447]: Fine as is, or revert to the index.

**Review by maxdiebold-erg — CHANGES_REQUESTED** [review]
Looks like you caught a lot of them!

I noticed that you used the `uuid` function to address the list index issues in SonarCloud. This might satisfy SonarCloud, but I think this would actually result in worse behavior performance-wise: the `uuid` is random by design, so this will create a new key on each render. Using an array index is as stable as the base array, which in many cases is set in state and is actually quite stable. Still, using the index is not recommended, so using some property or a deterministic hash of it should be used instead. I doubt it will affect performance in most cases, but it's something to keep in mind.

One simple hash (a Java built-in but taken from https://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript) could be implemented as:
```
function simpleHash(str) {
  let hash = 0;
  if (str.length === 0) return hash;

  for (let i = 0; i < str.length; i++) {
    const char = str.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash |= 0; // Convert to 32bit integer
  }

  return hash;
}
```
The other suggestion on the same SO post looks promising, too. I'll note the occurrences I find below.

**Comment by cschwinderg** [comment]
Some of them I left using `uuid`. I switched some back to `index` and just marked as a false positive in sonarcloud. 

I ended up not using `simpleHash` and instead just used an attribute (or combination of attributes) from the data being displayed. I feel like `simpleHash` doesn't really gain us much. It won't really enforce uniqueness, because the same string will result in the same hash. The only benefit I see is shorter keys in some instances, but I don't think that is worth the extra logic.

**Comment by maxdiebold-erg** [comment]
Sounds good. Right, the hash would only be for shorter keys, I used it more as an example as opposed to a random ID.

File: app/client/src/components/shared/AddSaveDataWidget.SearchPanel.tsx [hunk 8]
```diff
@@ -1,6 +1,6 @@
 /** @jsxImportSource @emotion/react */
 
-import {
+import React, {
```
Comment (maxdiebold-erg) [inline app/client/src/components/shared/AddSaveDataWidget.SearchPanel.tsx:3]: No issue here, but I tested this out and it looks like it doesn't need to be imported if it's only being used as a namespace for types.

**Review by maxdiebold-erg — APPROVED** [review]
Thanks for making those changes, I think the other changes look good.

**Pull request merged**